### PR TITLE
feat(mobile): Remote thumbnails and images use an on-disk image cache

### PR DESCRIPTION
**Pull request by martyfuhry**
- Fixed a few issues in the equals operator for thumbnails
- Replaced all references of `ImmichRemoteImageProvider` using `thumbnail: true` with `ImmichRemoteThumbnailProvider`
- Replaces the HTTP lookup with a `ImageCacheManager.getImageFile` which uses an on device cache
- Custom cache uses 5000 thumbnails and 500 "full size" remote images with a stale period of 30 days

### CODE REVIEW

File: mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import 'package:flutter_cache_manager/flutter_cache_manager.dart';
+import 'package:immich_mobile/shared/models/store.dart';
+
+/// Loads the codec from the URI and sends the events to the [chunkEvents] stream
+///
+/// Credit to [flutter_cached_network_image](https://github.com/Baseflow/flutter_cached_network_image/blob/develop/cached_network_image/lib/src/image_provider/_image_loader.dart)
+/// for this wonderful implementation of their image loader
+class ImageLoader {
+  static Future<ui.Codec> loadImageFromCache(
+    Uri uri, {
```
Comment (fyfrey) [inline mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart:14]: I think we can just use String.
All places in the code using this function can omit Uri.parse.

File: mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      headers: headers,
+      maxHeight: height,
+      maxWidth: width,
+    );
+
+    await for (final result in stream) {
+      if (result is DownloadProgress) {
+        // We are downloading the file, so update the [chunkEvents]
+        chunkEvents.add(
+          ImageChunkEvent(
```
Comment (fyfrey) [inline mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart:37]: Do we actually need these events? Afaik, we don't use them anywhere to show download progress? The loading spinners don't need the information

File: mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        // We have the file
+        final file = result.file;
+        final bytes = await file.readAsBytes();
+        final buffer = await ui.ImmutableBuffer.fromUint8List(bytes);
+        final decoded = await decode(buffer);
+        return decoded;
+      }
+    }
+
+    throw Exception('No file found in cache');
```
Comment (fyfrey) [inline mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart:54]: I hope we never reach this exception. If we do, it's a bit misleading. This exception can also occur if the download failed I believe.

File: mobile/lib/modules/asset_viewer/image_providers/cache/thumbnail_image_cache_manager.dart [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import 'package:flutter_cache_manager/flutter_cache_manager.dart';
+
+/// The cache manager for full size images [ImmichRemoteThumbnailProvider]
```
Comment (fyfrey) [inline mobile/lib/modules/asset_viewer/image_providers/cache/thumbnail_image_cache_manager.dart:3]: Minor typo in the comment: full size -> thumbnail

File: mobile/lib/modules/asset_viewer/image_providers/cache/remote_image_cache_manager.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+  factory RemoteImageCacheManager() {
+    return _instance;
+  }
+
+  RemoteImageCacheManager._()
+      : super(
+          Config(
+            key,
+            maxNrOfCacheObjects: 500,
```
Comment (fyfrey) [inline mobile/lib/modules/asset_viewer/image_providers/cache/remote_image_cache_manager.dart:16]: Can we limit the large remote image cache by total size instead of objects? Like 1GB etc.?

Anyhow, we can make this configurable in another PR

**Review by fyfrey — COMMENTED** [review]
Nice! This should improve the app experience substantially

File: mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart (hunk 2 above)
Comment (martyfuhry) [reply mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart:37]: We don't use it anywhere, and we wouldn't ever want to with thumbnails. With "full size" remote images? We may want to support this, but we do multi-stage loading... Hmm, I'm leaning towards tearing this out, or maybe making it optional if `chunkEvents` is sent in.

File: mobile/lib/modules/asset_viewer/image_providers/cache/remote_image_cache_manager.dart (hunk 5 above)
Comment (martyfuhry) [reply mobile/lib/modules/asset_viewer/image_providers/cache/remote_image_cache_manager.dart:16]: No, this is only configurable by max number of cache objects, not by size.

https://github.com/Baseflow/flutter_cache_manager/issues/265

File: mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart (hunk 2 above)
Comment (martyfuhry) [reply mobile/lib/modules/asset_viewer/image_providers/cache/image_loader.dart:37]: I made it optional and currently only `ImmichRemoteImageProvider` uses it. The gallery pager does use it, so we could optionally show some loading indicator in our placeholder for that.

**Review by fyfrey — APPROVED** [review]

**Review by alextran1502 — APPROVED** [review]

**Comment by alextran1502** [comment]
Amazing work! Thank you
